File: ts/engine/test/conformance/introspection.test.ts
import { readFileSync } from 'node:fs';
import { describe, expect, it, afterAll, beforeAll } from 'vitest';
import type { DbExecutor } from '../../src/index';
import { createPostgresExecutor } from '../../src/postgres/index';
import { createMysqlExecutor } from '../../src/mysql/index';
import { createSqliteExecutor } from '../../src/sqlite/index';
import { createMssqlExecutor } from '../../src/mssql/index';
import { introspectSchema, type SchemaData, type SchemaTable } from '../../src/introspection/index';

/**
 * Replays CORPUS D (introspection) against every dialect this engine can read.
 *
 * The corpus is AUTHORED from `harness/seed/*.sql` — the DDL is the source of truth, not whatever
 * these readers happen to return. A failure means the reader is wrong until proven otherwise, and
 * "proven otherwise" means an argument about what the catalog genuinely says, followed by an edit to
 * the corpus with the reason written down.
 *
 * Requires the shared docker harness (`pnpm harness:up`). FAIL LOUD, NEVER SKIP: a suite that skips
 * on an unreachable database reports green having tested nothing.
 */

type Dialect = 'mssql' | 'mysql' | 'postgres' | 'sqlite';
type Shared<T> = T | Record<Dialect, T>;
type CorpusColumn = {
  name: string;
  dataType: Shared<string>;
  nullable: Shared<boolean>;
  isPrimaryKey: Shared<boolean>;
  defaultValue: Shared<string | null>;
};
type CorpusForeignKey = {
  columnName: string;
  referencedTable: string;
  referencedColumn: string;
  referencedSchema: Shared<string | null>;
};
type CorpusIndex = { name: Shared<string | null>; columns: string[]; unique: boolean };
type CorpusTable = {
  name: string;
  type: 'table' | 'view';
  columns: CorpusColumn[];
  foreignKeys: CorpusForeignKey[];
  indexes: CorpusIndex[];
  approxRowCount: Shared<'present' | 'absent'>;
};
type Corpus = {
  version: string;
  schemaName: Record<Dialect, string>;
  tables: CorpusTable[];
};

const DIALECTS: Dialect[] = ['mssql', 'mysql', 'postgres', 'sqlite'];

const corpus = JSON.parse(
  readFileSync(
    new URL('../../../../contract/corpora/introspection/corpus.json', import.meta.url),
    'utf8',
  ),
) as Corpus;

/**
 * Resolve a corpus field for one dialect.
 *
 * A per-dialect map is a JSON object carrying EVERY dialect key; anything else is the shared value.
 * A partial map is a corpus bug, not a "use the shared value" shorthand — it is rejected loudly so a
 * half-written divergence cannot read as agreement.
 */
function pick<T>(field: Shared<T>, dialect: Dialect): T {
  if (field === null || typeof field !== 'object' || Array.isArray(field)) return field as T;
  const map = field as Record<string, T>;
  const missing = DIALECTS.filter((d) => !(d in map));
  if (missing.length > 0) {
    throw new Error(
      `corpus D: per-dialect map is missing ${missing.join(', ')} — spell out every dialect or use a shared value`,
    );
  }
  return map[dialect]!;
}

/** The corpus's expectation for one table, flattened to the dialect under test. */
function expected(table: CorpusTable, dialect: Dialect) {
  return {
    columns: table.columns.map((c) => ({
      name: c.name,
      dataType: pick(c.dataType, dialect),
      nullable: pick(c.nullable, dialect),
      isPrimaryKey: pick(c.isPrimaryKey, dialect),
      defaultValue: pick(c.defaultValue, dialect),
    })),
    foreignKeys: table.foreignKeys.map((fk) => ({
      columnName: fk.columnName,
      referencedTable: fk.referencedTable,
      referencedColumn: fk.referencedColumn,
      referencedSchema: pick(fk.referencedSchema, dialect),
    })),
    // A null name means the dialect reports no such index at all — drop it rather than expect one.
    indexes: table.indexes
      .map((i) => ({ name: pick(i.name, dialect), columns: i.columns, unique: i.unique }))
      .filter((i): i is { name: string; columns: string[]; unique: boolean } => i.name !== null),
    rowCount: pick(table.approxRowCount, dialect),
  };
}

/** The reader's answer for one table, in the same shape — `undefined` normalized to `null`. */
function actual(table: SchemaTable) {
  return {
    columns: table.columns.map((c) => ({
      name: c.name,
      dataType: c.dataType,
      nullable: c.nullable,
      isPrimaryKey: c.isPrimaryKey,
      defaultValue: c.defaultValue ?? null,
    })),
    foreignKeys: table.foreignKeys.map((fk) => ({
      columnName: fk.columnName,
      referencedTable: fk.referencedTable,
      referencedColumn: fk.referencedColumn,
      referencedSchema: fk.referencedSchema ?? null,
    })),
    indexes: table.indexes,
    rowCount: table.approxRowCount === undefined ? ('absent' as const) : ('present' as const),
  };
}

const byName = <T extends { name: string }>(items: T[]): T[] =>
  [...items].sort((a, b) => a.name.localeCompare(b.name));

describe('corpus D — schema introspection', () => {
  const executors: Record<string, DbExecutor> = {};
  const schemas: Record<string, SchemaData> = {};

  beforeAll(async () => {
    executors.postgres = createPostgresExecutor({
      host: 'localhost',
      database: 'sqleasy_ci',
      user: 'sqleasy',
      password: 'sqleasy_ci',
    });
    executors.mysql = createMysqlExecutor({
      host: 'localhost',
      database: 'sqleasy_ci',
      user: 'root',
      password: 'sqleasy_ci',
    });
    executors.mssql = createMssqlExecutor({
      server: 'localhost',
      database: 'sqleasy_ci',
      user: 'sa',
      password: 'SqlEasy_ci_1!',
      options: { encrypt: true, trustServerCertificate: true },
    });
    executors.sqlite = createSqliteExecutor({ url: ':memory:' });
    const seed = readFileSync(
      new URL('../../../../harness/seed/sqlite.sql', import.meta.url),
      'utf8',
    )
      .split('\n')
      .filter((line) => !line.trim().startsWith('--'))
      .join('\n')
      .split(';')
      .map((s) => s.trim())
      .filter(Boolean);
    for (const statement of seed) await executors.sqlite.run({ sql: statement });

    for (const dialect of DIALECTS) {
      schemas[dialect] = await introspectSchema(
        executors[dialect]!,
        dialect,
        dialect === 'sqlite' ? undefined : corpus.schemaName[dialect],
      );
    }
  }, 30_000);

  afterAll(async () => {
    for (const executor of Object.values(executors)) await executor.close();
  });

  it('replays every dialect the engine can read', () => {
    expect(Object.keys(executors).sort()).toEqual([...DIALECTS].sort());
  });

  for (const dialect of DIALECTS) {
    describe(dialect, () => {
      it('reports exactly the seeded tables and views', () => {
        expect(byName(schemas[dialect]!.tables).map((t) => `${t.type} ${t.name}`)).toEqual(
          byName(corpus.tables).map((t) => `${t.type} ${t.name}`),
        );
      });

      it('scopes every table to the dialect’s own namespace', () => {
        const namespaces = [...new Set(schemas[dialect]!.tables.map((t) => t.schema))];
        expect(namespaces).toEqual([corpus.schemaName[dialect]]);
      });

      for (const table of corpus.tables) {
        it(`${table.name}: columns, keys, indexes, row-count presence`, () => {
          const found = schemas[dialect]!.tables.find((t) => t.name === table.name);
          expect(found, `${dialect} has no table named ${table.name}`).toBeDefined();
          const want = expected(table, dialect);
          const got = actual(found!);

          // Columns in ordinal order — position IS the contract.
          expect(got.columns).toEqual(want.columns);
          expect(byName(got.foreignKeys.map((f) => ({ ...f, name: f.columnName })))).toEqual(
            byName(want.foreignKeys.map((f) => ({ ...f, name: f.columnName }))),
          );
          // Indexes as a set: enumeration order is the reader's ORDER BY, not the catalog's promise.
          expect(byName(got.indexes)).toEqual(byName(want.indexes));
          expect(got.rowCount).toBe(want.rowCount);
        });
      }
    });
  }
});

File: dart/engine/test/conformance/introspection_test.dart
@TestOn('vm')
@Tags(['integration'])
library;

import 'dart:convert';
import 'dart:io';

import 'package:sqleasy_engine/sqleasy_engine.dart';
import 'package:test/test.dart';

/// Replays CORPUS D (schema introspection) against every dialect this port can read.
///
/// The corpus — not this engine — is the definition of correct: it is AUTHORED from
/// `harness/seed/*.sql`, so a failure means the reader is wrong until an argument about what the
/// catalog genuinely says proves otherwise.
///
/// Requires the shared docker harness (`pnpm harness:up`). FAIL LOUD, NEVER SKIP.
void main() {
  // MSSQL is absent from this port ENTIRELY, not merely unreplayed: there is no pure-Dart TDS
  // driver, so `IntrospectDialect` has no `mssql` value to pass. The TypeScript replay covers all
  // four. This set is asserted below so the gap can never go quiet.
  const implemented = <String, IntrospectDialect>{
    'postgres': IntrospectDialect.postgres,
    'mysql': IntrospectDialect.mysql,
    'sqlite': IntrospectDialect.sqlite,
  };

  final corpus = jsonDecode(
    File('../../contract/corpora/introspection/corpus.json').readAsStringSync(),
  ) as Map<String, Object?>;
  final schemaName = (corpus['schemaName']! as Map).cast<String, String>();
  final tables = (corpus['tables']! as List).cast<Map<String, Object?>>();

  /// Resolve a corpus field for one dialect.
  ///
  /// A per-dialect map is a JSON object naming EVERY dialect the contract knows — including `mssql`,
  /// which this port cannot read but the corpus still speaks for. Anything else is the shared value.
  /// A partial map is a corpus bug, never a "fall back to shared" shorthand, so it fails loudly
  /// rather than letting a half-written divergence read as agreement.
  Object? pick(Object? field, String dialect) {
    if (field is! Map) return field;
    const known = {'mssql', 'mysql', 'postgres', 'sqlite'};
    final missing = known.difference(field.keys.cast<String>().toSet());
    if (missing.isNotEmpty) {
      fail('corpus D: per-dialect map is missing ${missing.join(', ')} — '
          'spell out every dialect or use a shared value');
    }
    return field[dialect];
  }

  late Map<String, DbExecutor> executors;
  late Map<String, SchemaData> schemas;

  setUpAll(() async {
    executors = {
      'postgres': await PostgresExecutor.open(
        const PostgresConnectionOptions(
          host: 'localhost',
          database: 'sqleasy_ci',
          username: 'sqleasy',
          password: 'sqleasy_ci',
        ),
      ),
      'mysql': await MysqlExecutor.open(
        const MysqlConnectionOptions(
          host: 'localhost',
          database: 'sqleasy_ci',
          username: 'root',
          password: 'sqleasy_ci',
          allowInvalidCertificate: true,
        ),
      ),
      'sqlite': SqliteExecutor.openInMemory()
        ..database.execute(
          File('../../harness/seed/sqlite.sql').readAsStringSync(),
        ),
    };
    schemas = {
      for (final entry in implemented.entries)
        entry.key: await introspectSchema(
          executors[entry.key]!,
          entry.value,
          entry.key == 'sqlite' ? null : schemaName[entry.key],
        ),
    };
  });

  tearDownAll(() async {
    for (final executor in executors.values) {
      await executor.close();
    }
  });

  test('replays every dialect this port can read', () {
    expect(executors.keys.toSet(), implemented.keys.toSet());
    expect(IntrospectDialect.values.map((d) => d.name).toSet(),
        implemented.keys.toSet(),
        reason:
            'a dialect the port can read but does not replay is a coverage hole');
  });

  for (final dialect in implemented.keys) {
    group(dialect, () {
      test('reports exactly the seeded tables and views', () {
        final got = schemas[dialect]!
            .tables
            .map((t) => '${t.type.name} ${t.name}')
            .toList()
          ..sort();
        final want = tables.map((t) => '${t['type']} ${t['name']}').toList()
          ..sort();
        expect(got, want);
      });

      test('scopes every table to the dialect’s own namespace', () {
        expect(schemas[dialect]!.tables.map((t) => t.schema).toSet(),
            {schemaName[dialect]});
      });

      for (final table in tables) {
        final name = table['name']! as String;
        test('$name: columns, keys, indexes, row-count presence', () {
          final found =
              schemas[dialect]!.tables.where((t) => t.name == name).toList();
          expect(found, hasLength(1),
              reason: '$dialect has no table named $name');
          final got = found.single;

          // Columns in ordinal order — position IS the contract.
          final wantColumns = (table['columns']! as List)
              .cast<Map<String, Object?>>()
              .map((c) => [
                    c['name'],
                    pick(c['dataType'], dialect),
                    pick(c['nullable'], dialect),
                    pick(c['isPrimaryKey'], dialect),
                    pick(c['defaultValue'], dialect),
                  ])
              .toList();
          expect(
            got.columns
                .map((c) => [
                      c.name,
                      c.dataType,
                      c.nullable,
                      c.isPrimaryKey,
                      c.defaultValue,
                    ])
                .toList(),
            wantColumns,
          );

          final wantFks = (table['foreignKeys']! as List)
              .cast<Map<String, Object?>>()
              .map((f) => [
                    f['columnName'],
                    f['referencedTable'],
                    f['referencedColumn'],
                    pick(f['referencedSchema'], dialect),
                  ])
              .toList()
            ..sort((a, b) => '${a.first}'.compareTo('${b.first}'));
          expect(
            got.foreignKeys
                .map((f) => [
                      f.columnName,
                      f.referencedTable,
                      f.referencedColumn,
                      f.referencedSchema,
                    ])
                .toList()
              ..sort((a, b) => '${a.first}'.compareTo('${b.first}')),
            wantFks,
          );

          // Indexes as a SET: enumeration order is the reader's ORDER BY, not the catalog's promise.
          // A null name means the dialect reports no such index at all — drop it, don't expect one.
          final wantIndexes = (table['indexes']! as List)
              .cast<Map<String, Object?>>()
              .map((i) => [pick(i['name'], dialect), i['columns'], i['unique']])
              .where((i) => i.first != null)
              .toList()
            ..sort((a, b) => '${a.first}'.compareTo('${b.first}'));
          expect(
            got.indexes.map((i) => [i.name, i.columns, i.unique]).toList()
              ..sort((a, b) => '${a.first}'.compareTo('${b.first}')),
            wantIndexes,
          );

          // A PRESENCE flag, never a number: it comes from planner statistics that are explicitly
          // allowed to be stale (Postgres reports -1 until ANALYZE; SQLite has no such statistic).
          expect(got.approxRowCount == null ? 'absent' : 'present',
              pick(table['approxRowCount'], dialect));
        });
      }
    });
  }
}
